common/decompressingWriter.go

```diff
@@ -0,0 +1,138 @@
+// Copyright © Microsoft <wastore@microsoft.com>
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package common
+
+import (
+	"compress/gzip"
+	"compress/zlib"
+	"errors"
+	"io"
+	"time"
+)
+
+type decompressingWriter struct {
+	pipeWriter  *io.PipeWriter
+	workerError chan error
+}
+
+const decompressingWriterCopyBufferSize = 256 * 1024 // 1/4 the size that we usually write to disk with (elsewhere in codebase). 1/4 to try to keep mem usage a bit lower, without going so small as to compromize perf
```

Comment:
Do you think this would cause a bottleneck? Thinking about it, we'd never actually fill up the larger buffer in theory.

Reply:
I wasn't so much worried about perf, but rather about the fact that these buffers aren't tracked within our memory budget (that we use for the main chunk buffers).  So if there are a lot of compressed files getting decompressed in parallel - e.g. a thousand (which is possible if they are small) and concurrency is high - then that starts to add up to some non-trivial mem usage.  Not sure if I understood your question properly tho...

Reply:
I understand that much. Perhaps we could make it so that the buffers are split half and half when decompression is enabled.

My thought is mostly that we're trying to squeeze data from a smaller pipe into a larger pipe-- we'd never actually fill the larger pipe up, and our throughput will be lowered by virtue of this.

Reply:
>My thought is mostly that we're trying to squeeze data from a smaller pipe into a larger pipe-- we'd never actually fill the larger pipe up, and our throughput will be lowered by virtue of this

I'm not sure I've understood that, can you elaborate?

Reply:
BTW, it's also worth noting that they copyBuffer call that uses this buffer just keeps iterating through the data until its all done. So the only ways this could hurt perf are:
1. The CPU cost of the extra iterations (won't be an issue because it's virtually always disk or network or Service that's the bottleneck, not CPU)
2. A change in disk write performance due to the smaller buffer size.  This is the one I'm slightly concerned about. But there's probably a wide sweet-spot of suitable values, and I doubt this size pulls us out of there.  E.g. using something like 32KB might change the perf considerably, but dropping from 1 MB to 1/4 MB as we are doing here is unlikely, as far as I can tell, to make any significant change.
3. Using too much RAM, and causing paging.  This is why I made it 1/4 of what we normally use.

I haven't fully thought through the implications of trying to use the same pool as we do for the main chunk buffers. I fear there could be potential for starvation or deadlock there - e.g. all our chunk buffers are allocated, but they can't be saved because we can't allocate any decompression buffers because the chunk buffers are taking up all the RAM.  

Given all of the above, and the need to keep our workflow as simple as possible this week in terms of getting PRs completed, I propose to do no changes to this code in this PR.  Please let me know if you think we should create a new backlog item, and or PR, to test and tweak this buffer size.